File: solutions/_1849_splitting_a_string_into_descending_consecutive_values/solution.py
from functools import cache


class Solution:
    def splitString(self, s: str) -> bool:
        @cache
        def split(i, prev, count):
            if i >= len(s):
                return count >= 2
            num = 0
            for j in range(i, len(s)):
                num = num * 10 + int(s[j])
                if prev == -1 or num == prev - 1:
                    if split(j + 1, num, count + 1):
                        return True
            return False

        return split(0, -1, 0)


# https://leetcode.com/problems/splitting-a-string-into-descending-consecutive-values/

File: solutions/_1849_splitting_a_string_into_descending_consecutive_values/solution.cpp

#include <vector>
#include <string>
using namespace std;

class Solution
{
private:
    string s;

    bool split(int i, long long prev, int count)
    {
        if (i >= s.size())
            return count >= 2;

        unsigned long long num = 0;
        for (int j = i; j < s.size(); j++)
        {
            num = num * 10 + s[j] - '0';
            if (prev == -1 || prev - 1 == num)
                if (split(j + 1, num, count + 1))
                    return true;
        }

        return false;
    }

public:
    bool splitString(string s)
    {
        this->s = s;
        return split(0, -1, 0);
    }
};

/*
https://leetcode.com/problems/splitting-a-string-into-descending-consecutive-values/
*/